cpp/src/io/parquet/predicate_pushdown.cpp

```diff
@@ -0,0 +1,501 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "reader_impl_helpers.hpp"
+
+#include <cudf/ast/detail/expression_parser.hpp>  // possibly_null_value defined here, but used in operators.hpp
+#include <cudf/ast/detail/expression_transformer.hpp>
+#include <cudf/ast/detail/operators.hpp>
+#include <cudf/ast/expressions.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/utilities/integer_utils.hpp>
+#include <cudf/detail/utilities/vector_factories.hpp>
+#include <cudf/transform.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/mr/device/per_device_resource.hpp>
+
+#include <algorithm>
+#include <list>
+#include <optional>
+
+namespace cudf::io::detail::parquet {
+
+/**
+ * @brief Converts statistics in column chunks to 2 device columns - min, max values.
+ *
+ */
+struct stats_caster {
+  size_type total_row_groups;
+  std::vector<metadata> const& per_file_metadata;
+  host_span<std::vector<size_type> const> row_group_indices;
+  rmm::cuda_stream_view stream;
+
+  template <typename ToType, typename FromType>
+  static ToType targetType(FromType const value)
+  {
+    if constexpr (cudf::is_timestamp<ToType>()) {
+      return static_cast<ToType>(
+        typename ToType::duration{static_cast<typename ToType::rep>(value)});
+    } else if constexpr (std::is_same_v<ToType, string_view>) {
+      return ToType{nullptr, 0};
+    } else {
+      return static_cast<ToType>(value);
+    }
+  }
+  // uses storage type as T
+  template <typename T>
+  static T convert(uint8_t const* stats_val, size_t stats_size, cudf::io::parquet::Type const type)
+  {
+    switch (type) {
+      case BOOLEAN: return targetType<T>(*reinterpret_cast<bool const*>(stats_val));
+      case INT32: return targetType<T>(*reinterpret_cast<int32_t const*>(stats_val));
+      case INT64: return targetType<T>(*reinterpret_cast<int64_t const*>(stats_val));
+      case INT96:  // Deprecated
+        return targetType<T>(static_cast<__int128_t>(reinterpret_cast<int64_t const*>(stats_val)[0])
+                               << 32 |
+                             reinterpret_cast<int32_t const*>(stats_val)[2]);
+      case FLOAT:
+        if constexpr (std::is_floating_point_v<T>)
```

Comment:
This check isn't included on the integer or boolean values. Why do we need to check it here? If the caller asks for a float out of a bool pointer it would be undefined, right? If it is an alignment issue, why not check the int64?

Reply:
moved to type specializations for different types.